### ISSUE TITLE
Attempt to Build IOS Xcode Project/Android APK Fails

### ISSUE THREAD

**Issue by AndrewKahr** [PlayEveryWare/eos_plugin_for_unity#149]
**Describe the bug**
When attempting to build an Xcode project targeting IOS or an Android APK, the build fails with the error `Library\PackageCache\com.playeveryware.eos@85253aec745c\Runtime\DLLHandle.cs(218,51): error CS1061: 'SystemDynamicLibrary' does not contain a definition for 'UnloadLibrary' and no accessible extension method 'UnloadLibrary' accepting a first argument of type 'SystemDynamicLibrary' could be found (are you missing a using directive or an assembly reference?)`.

**To Reproduce**
1. Build the EOS Unity plugin from the release-2.1.5 branch
2. Install built tarball using Package Manager
3. Import Example Scenes
4. Configure EOS config with IDs and Keys
5. Switch platform to IOS or Android
6. Build the project

**Expected behavior**
Build succeeds

**Screenshots**
![image](https://user-images.githubusercontent.com/22359829/196013625-faa8e4a9-d064-4265-a1cb-af82e1e3d309.png)

**Desktop (please complete the following information):**
 - OS: Windows 10 21H2 x64
 - Unity Version: 2022.2.0b10
 - Plugin Version: 2.1.5
 
**Additional context**
Able to play with the backend set to IOS/Android in the editor because the code throwing the compile error is only enabled via ifdef when the backend isn't `UNITY_EDITOR`.

**Comment by andrew-hirata-playeveryware** [comment]
Hmm A quick way to get this to work, since you're want Android and iOS support, is to enable `EOS_PREVIEW_PLATFORM` . That being said, by default this was intended to still work.

**Comment by AndrewKahr** [comment]
That got us a step further! Was able to build an APK and an xcode project. Unfortunately, running into this error when attempting to compile the xcode  project:
![image](https://user-images.githubusercontent.com/22359829/196261429-889d0f88-054d-4265-995f-165f3f27b627.png)
Is there anything else that needs to be enabled?
Really appreciate all the help you've been providing!

**Comment by andrew-hirata-playeveryware** [comment]
Hmm as a quick fix, I think if you copy over the [Objective-C++ files & meta files](https://github.com/PlayEveryWare/eos_plugin_for_unity/tree/stable/Assets/Plugins/iOS) from the iOS plugin directory to somewhere in your project's Assets/Plugin/iOS directory , you might be able to get past those build errors. Long term I'll need to double check that things are being packaged correctly for iOS and that they build correctly in a standalone build, because I somehow _wasn't_ seeing this when I tried to test on my local mac mini.

**Comment by AndrewKahr** [comment]
One step closer now! I got it to build however when running it throws these errors:
`
<color=#FF0000>20:29:30.456 [Error] : Platform not supported</color>
20:29:30.475 [Log] : ConfigureSystemInitOptions
<color=#FF0000>20:29:30.478 [Exception] : FileNotFoundException: Could not find file "/private/var/containers/Bundle/Application/C3DCD4D8-DFA8-4C5A-BD75-3BD32037EDCE/App.app/Data/Raw/EOS/eos_ios_config.json"</color>
20:34:01.501 [Log] : Attempting to login...
<color=#FF0000>20:34:01.507 [Exception] : NullReferenceException: Object reference not set to an instance of an object.</color>
`
I notice it's looking for `eos_ios_config.json` but I only have `EpicOnlineServices.json` generated in StreamingAssets. Was hoping it was an innocuous error hence the attempt at logging in however it seems to be a blocking issue. Also the very first line is saying it isn't a supported platform, not sure if there's another define I need to fix that.

Some notes on the test run, the errors before `Attempting to login` were on immediate launch.

**Comment by AndrewKahr** [comment]
Just tried adding an override for ios config which generated the json and got it to run a bit further. Now it is throwing:
```
20:56:38.565 [Error] : Platform not supported
20:56:38.585 [Log] : ConfigureSystemInitOptions
20:56:38.586 [Log] : GetEOS_iOSConfig() is not null
20:56:38.587 [Log] : EOSManager: Registering for platform-specific notifications
20:56:38.592 [Exception] : DllNotFoundException: Unable to load DLL 'EOSSDK.framework/EOSSDK'. Tried the load the following dynamic libraries: Unable to load dynamic library '/EOSSDK.framework/EOSSDK' because of 'Failed to open the requested dynamic library (0x06000000) dlerror() = dlopen(/EOSSDK.framework/EOSSDK, 0x0005): tried: '/EOSSDK.framework/EOSSDK' (no such file)
```
I suspect I also need to copy framework into my plugins folder.

**Comment by AndrewKahr** [comment]
That indeed seemed to do the trick. I think looking at the package description, it needs lines to copy the framework and mm files as it currently only grabs the cs files. 

Now that it's actually launched, there appears to be a new problem with oauth. On desktop it would open the browser to link the client permissions but on mobile it seems to just fail:
```
21:07:33.484 [Error] : Platform not supported
21:07:33.504 [Log] : ConfigureSystemInitOptions
21:07:33.504 [Log] : GetEOS_iOSConfig() is not null
21:07:33.505 [Log] : EOSManager: Registering for platform-specific notifications
21:07:33.517 [Warning] : EOSManager::Init: InitializePlatformInterface: initResult = Success
21:07:33.521 [Log] : c7c3d81b1ac9492fb8c341b318fdd6c5
21:07:33.711 [Warning] : 2022-10-17T14:07:33.5306100-07:00 LogEOSUI(Warning): EOS_UI_AddNotifyDisplaySettingsUpdated failed because the overlay is not currently configured.
21:07:34.301 [Warning] : 2022-10-17T14:07:34.3010760-07:00 LogEOSAnalytics(Warning): EOS SDK Analytics disabled for route [1].
21:07:34.663 [Warning] : 2022-10-17T14:07:34.6632680-07:00 LogEOSEcom(Warning): Purchase flow is disabled due to overlay setup failure (EOS_NotConfigured).
21:08:04.875 [Log] : Attempting to login...
21:08:04.977 [Warning] : 2022-10-17T14:08:04.9774870-07:00 LogEOS(Warning): Error response received from backend. ServiceName=[OAuth], OperationName=[TokenGrantv2], Url=[<Redacted>], HttpStatus=[400], ErrorCode=[errors.com.epicgames.common.oauth.invalid_request], NumericErrorCode=[1013], ErrorMessage=[exchange_code is required.], CorrId=[EOS-Cd76qagVTDylKJiOXdrOAQ-I3JwljrJREy9SIf4U2lAEg-_muZSpl2R-qPEb267tOBQw]
21:08:05.079 [Log] : Error logging in. [InvalidRequest]
```
I'm testing with the playerdata sample scene with the dev auth tool running on my local network with a firewall rule configured to allow it. I had this warning when running on desktop too but it would let me log in after that initial auth screen on a browser.

**Comment by AndrewKahr** [comment]
Just had a chance to test on Android and got the following error log (Apologies for the long read ahead):
```
07:31:09.047 [Error] : Platform not supported
07:31:09.058 [Log] : EOSAndroid: Getting activity context...
07:31:09.073 [Log] : EOSAndroid: activity context found!
07:31:09.077 [Exception] : AndroidJavaException: java.lang.ClassNotFoundException: com.epicgames.mobile.eossdk.EOSSDK
```
Which I assumed was because it had the same issue as the iOS side, so I copied over the `eos_dependencies.androidlib` folder, `DynamicLibraryLoaderHelper_Android.cpp, eos-sdk.aar, gradleTemplate.properties, mainTemplate.gradle, proguard-user.txt, UnityHelpers_android-debug.aar, and all associated meta files. 

When trying to build, I get a popup window that says `mainTemplate.gradle file is using the old aaptOptions noCompress property definition which does not include types defined by unityStreamingAssets constant.` [Googling](https://support.unity.com/hc/en-us/articles/4408584577044-Why-do-I-get-errors-when-using-a-Gradle-file-with-an-old-aaptOptions-noCompress-property-) seemed to point to modifying `mainTemplate.gradle` under `aaptOptions` to add `noCompress = ['.ress', '.resource', '.obb'] + unityStreamingAssets.tokenize(', ')`. Unity then asked to automatically update the file and added `unityStreamingAssets=.unity3d**STREAMING_ASSETS**` to `gradleTemplate.properties`. Unity tried to build again but ran into an error saying `android.enableR8` needed to be either true or false. I ended up just removing the line `android.enableR8=**MINIFY_WITH_R_EIGHT**` from `gradleTemplate.properties`. Next it threw the error 
```
> Failed to install the following Android SDK packages as some licences have not been accepted.
     build-tools;30.0.3 Android SDK Build-Tools 30.0.3
     platforms;android-29 Android SDK Platform 29
```
I think this is caused by using Unity 2022.2 as they are migrating android tools to 32 and the JDK up to 11 as noted [here](https://forum.unity.com/threads/android-tools-in-unity-2022-2.1294815/). I ended up changing 
```
compileSdkVersion 29
    buildToolsVersion '30.0.3'
    defaultConfig {
        targetSdkVersion 29
    }
```
to
```
compileSdkVersion 32
    buildToolsVersion '32.0.0'
    defaultConfig {
        targetSdkVersion 32
    }
```
It then threw an error saying `minSdkVersion 22 cannot be smaller than version 23 declared in library eos-sdk` so I bumped the project target SDK version to 23. This ultimately allowed me to build an apk, however, I haven't been able to test on a physical device yet. I tried using bluestacks and it just gave a black screen after the unity logo while other projects built with 2022.2.0b10 play normally so I suspect there's something still wrong with the build. Will report back once I've tested it on a real device.

TLDR: I had to change several fields throughout the gradle files and project settings to get it through a bunch of build errors however it still doesn't seem to build properly despite outputting an apk. Also not entirely sure if the changes I made are valid or if they just happened to work (Not super familiar with Android stuff).

Is there anything I missed that made this process so involved or is this due to using Unity 2022.2.0b10 instead of Unity 2020? Also, should I start breaking these various errors into separate issues, or is this one thread better for issue tracking?

**Comment by andrew-hirata-playeveryware** [comment]
'Supported' means that we officially support it and the goal is that released versions of a few bugs as possible on that platform. IT also means that if something breaks it, we will try to do a hotfix to resolve the issue. iOS and Android are 'Preview' platforms, which means they're in the UPM, but there _will_ be bugs as we work to make things function smoothly, and we won't necessarily make hotfixes to resolve issues. And yes, some of those Android issues might be due to using a newer version of Unity. Other parts of it look like files are either not being used correctly from the UPM or are missing.

**Comment by AndrewKahr** [comment]
Completely understand on the platforms being in preview! I should note I never came with the expectation for this to work out of the box. My objective of the messages in this thread is to provide as much context as possible for what steps I am taking to try to get it to run so that others who may run into similar issues while the platforms are in preview have something to compare against and so you and other maintainers can more easily see what issues are being run into and how to reproduce them 🙂. 

My sincerest apologies if my messages came across as if there was an expectation these should behave like stable platforms! I really do appreciate all the support you've been able to provide so far despite these being preview platforms! And if there's anything I can do to make these issues more informative or useful for maintainers, please do let me know.

**Comment by AndrewKahr** [comment]
Bump not resolved

**Comment by andrew-hirata-playeveryware** [comment]
Just an update: we still need to address this with a fix that supports multiple android versions / unity versions.

**Comment by AndrewKahr** [comment]
Bump not resolved

**Comment by AndrewKahr** [comment]
After testing IOS more, it seems logging in via account portal with a real account works, it's just the dev auth that doesn't work and fails with the `exchange_code is required` error.

**Comment by andrew-hirata-playeveryware** [comment]
Sorry for the delay: I have some thoughts && updates:

Hmm I wonder if that's a change in account behavior or requirements on the EOS backend? Furthermore, I have observed some platforms not working with the dev auth tool, which is not a tool we (PlayEveryWare) control. I also recall that on iOS, one has to declare some network things one uses, so maybe that's an issue: i.e. iOS is blocking some port that dev auth uses, maybe? I usually use the account portal and persistent auth to login; I'd have to run some tests with dev auth and double check our sample code to see if it's broken in our samples or if our sample project is doing something special.

Concerning Android: I have some stuff I've been working on to remove some of the need to even package the gradle stuff, and instead install a copy of the eos_dependencies, which could then be modified by the user as they need.

**Comment by AndrewKahr** [comment]
Thanks for the update!

It's no worries if the dev auth tool doesn't work. The only reason I was using it was because the EOS documentation seemed to indicate you should use that during development; I didn't realize I could just use regular credentials instead with no issue. The idea that ios is blocking the port the auth tool uses is actually very possible as I know in newer iOS versions, apps need to request permissions for discovering local devices on the network which perhaps also includes servers like the auth tool.

On the Android side, I just had a chance to test on a real device. For some reason when using bluestacks to test, it was just giving a black screen after the Unity logo, but disabling frame pacing fixed the app load. However, in both bluestacks and the real Android device, clicking on login for the account portal doesn't do anything. It just logs a message saying "Attempt to login" and the login button counts down without launching any prompts or any other log messages. Is this something anyone else has experienced when testing Android? I can't tell if it is just a bug with the implementation or whether the new Unity version with new JDK and Android SDK/NDK versions are breaking it somehow.

It's also good to hear that it might be possible for Android dependencies to be decoupled from the package itself as during my testing to build an apk, the gradle file needed several modifications to work on Unity 2022.2.0b10+ and the androidlib within the Android runtime folder had additional hardcoded dependencies that needed to be bumped to newer versions as they all had hardcoded version numbers. Unfortunately, the Unity macros for Android versions didn't seem to work within the androidlib's gradle file.

It seems like it's really close to working on mobile though! Once I get past logging in, the samples seem to work flawlessly on ios which in theory should be the same for Android. At this stage, the only thing really blocking me would be the Android login and the modifications needed for gradle to function properly on Unity 2022.2.0+. Once again, really appreciate the work being done to build out the plugin and reach mobile support!

**Comment by andrew-hirata-playeveryware** [comment]
Yeah, that's what my memory tells me about iOS: I've added a separate task on our own backlog to get that working.

As far as the black screen goes on Android, I think I have a resolution for that, as apart of the work I did to remove the dependency on gradle template files. I believe that was caused by a hardcoded string value that needed to be changed so that it matches the one configured in the EpicOnlineConfigurationTool. 

Yeah, the Unity macros don't work in gradle because it treats them like normal groovy files, I believe. Luckily, it was quite easy to remove the dependency from the package and that change is `development`, and will be in the next release.

**Comment by noahbare** [comment]
> 

Dumb question, how do you enable EOS_PREVIEW_PLATFORM? I've searched the repo and can't find where that's assigned and I don't see that as an option anywhere in the editor.

**Comment by AndrewKahr** [comment]
If you go to your project settings and go to `player->other settings->scripting define symbols` you can add it there.

**Comment by AndrewKahr** [comment]
Bump need to validate the latest updates have resolved the issue

**Comment by AndrewKahr** [comment]
Was going to test it today actually, will report back soon.

**Comment by AndrewKahr** [comment]
@andrew-hirata-playeveryware Tried to test it but #170 is blocking me unfortunately.

**Comment by AndrewKahr** [comment]
Fixed the block in #173. When testing the Android build, however, I still see hardcoded values for the build.gradle file which causes a build error. Can these be dynamically filled with Unity's macros? This issue is caused by Unity upgrading their Android tools to 32 while the hardcoded version is 29 and 30 (which is the version that Unity used prior to 2022.2.X)

**Comment by AndrewKahr** [comment]
Bump not resolved with latest develop

**Comment by scovertb** [comment]
Currently working on substituting those values based on Unity version and project settings. Where did you find documentation of which Android build tools version Unity is using? I haven't been able to find it.

**Comment by scovertb** [comment]
In that case, the best I'll probably be able to do is select the highest 30.x.x build tools if lower than Unity 2022.2 and 32.x.x if it's higher.

**Comment by scovertb** [comment]
Is the Java version still an issue? I haven't been able to find any of our config templates that reference Java 8.

**Comment by AndrewKahr** [comment]
Just curious, why can't you just use the macros that Unity usually leverages for the gradle build? It sounds like you're planning on hard coding some values. 

And for Java version, I also don't see it referenced anywhere so it's likely a non issue. Unsure whether the system checks Java or android tool versions first so I'm unsure whether it's fully resolved or not.

**Comment by scovertb** [comment]
It seems like the macros are only applied to the specific files that Unity generates when enabling custom gradle configs, and there doesn't appear to be an editor scripting functions that replicate that functionality.

**Comment by scovertb** [comment]
If the build tool license is the main issue, you can try launching Android Studio and pointing the SDK Manager at the Unity Android SDK location, then installing any tools listed at "partially installed." That should give you a chance to accept any licenses and it should install the v30 build tools.

**Comment by AndrewKahr** [comment]
The tricky part is this also needs to happen in a CI system (gameCI specifically). So ideally we would just have the correct version without needing to accept new licenses. If the macro doesn't work, can we instead have a defines file using the `UNITY_2022_2_OR_NEWER` style defines to set variables for the build tools versions and a script to do the replacement before the build?

**Comment by scovertb** [comment]
Sorry, that was just a temporary suggestion. I should be able to get in macro replacement by the next release.

**Comment by AndrewKahr** [comment]
Sounds good, thank you so much!
